zip-bolt-support.rst

```diff
@@ -0,0 +1,408 @@
+   - ``outCiphertext``: encrypted output note (part 2)
+   - ``zkproof``: zero-knowledge proof for the note
+
+The merchant can initiate closing by posting the initial commitment transaction (in Section 2.3) from establishing the channel that pays the merchant the full balance of the channel with a time lock that allows for customer dispute. The merchant can then post a separate standard transaction that moves those funds to a shielded address.
+
+3. Custom Shielded Tx: Using Z-addresses and Scriptless
+-------------
+We assume the following features are present:
+
+(a) ``lock_time`` - for absolute lock time
+(b) A way to enforce relative lock time
+(c) 2-of-2 multi-sig shielded address support
+(d) All inputs/outputs are specified from/to a shielded address
+(e) A method to encumber the outputs of a shielded transaction
+(f) An extension to the transaction format to include BOLT (e.g., like ``vBoltDescription``)
+(g) Extend the ``SIGHASH`` flags to cover the extended field
+
+The goal here is to perform all the same validation steps for channel opening/closing without relying on the scripting system, as well as allowing for relative timelocks (the equivalent of ``OP_CSV``). In order to support multihop payments, we need absolute timelocks as well (the equivalent of ``OP_CLTV``). We also want to ensure that transactions are non-malleable in order to allow for unconfirmed dependency transaction chains.
+
+**Limitations/Notes**: With extensions to shielded transaction format, it may be evident whenever parties are establishing private payment channels. We appreciate feedback on the feasibility of what is proposed for each aspect of the Bolt protocol.
+
+3.1 Channel Opening
+-------------
+The customer creates a funding transaction that spends ZEC from a shielded address to a 2-of-2 multi-sig shielded address. Here is the flow (1) creating a multisig shielded address specifying both parties keys and (2) generating channel tokens.
+
+3.2 Funding Transaction
+-------------
+The funding transaction is by default funded by only one participant, the customer. It could also be funded by the merchant. 
+
+This transaction has 2 shielded inputs (but can be up to some N) and 1 output to a 2-of-2 shielded address with the merchant public key. If an ``vBoltDescription`` field is added, then we could use it to store the channel parameters and the channel token for opening the channel.
+
+3.3 Initial Wallet Commitment
+-------------
+The initial wallet commitment will spend from the shielded address to two outputs: a P2SH output (for customer) and P2PKH (for merchant).  The first output pays the customer with a timelock (or the merchant with a revocation token) and the second output allows the merchant to spend immediately. It is not clear to us whether it will be possible to encumber the outputs of shielded outputs directly. 
+
+We would appreciate feedback on the possibilities with creating commitment transactions via shielded transactions only.
```

Comment:
I think the simplest strategy would be to create a specific Bolt circuit that enforces the necessary encumberance, and then require that the spender provide a proof that takes in public inputs satisfying the encumberance. This could be implemented as a special case of P2VK (https://github.com/zcash/zcash/issues/2425), or a dedicated circuit that could be migrated in a future NU to a more general programmability solution  (general P2VK, or something else).

Broadly, I would imagine something along these lines:
- The encumbered output would contain a commitment to the various Bolt parameters (the timelock, the revocation token, etc).
  - Without changing the Sapling circuit, the commitment would be added to a global Merkle tree in parallel to the current Sapling Merkle tree (meaning that they don't have a shared privacy set).
  - If the Sapling circuit was altered, the privacy sets could potentially be shared, at the cost of requiring all Sapling users to be aware of Bolt semantics. IMHO this probably isn't worth the cost of doing such a change, but we could consider it during a later general programmability solution.
- The spend using that output would contain a proof using the Bolt circuit, and the necessary public inputs such as the "time" at which the proof was created (perhaps stored in the `locktime` field).
  - The circuit would enforce the equivalent of the `OP_BOLT` logic, allowing a valid proof to be created if the prover had knowledge of the revocation key and merchant key, OR the prover had knowledge of the customer key AND the public time input was past the committed timelock.
  - Network nodes would validate the Bolt-specific proof, and also validate the public inputs (if necessary, e.g. the `locktime` field is already enforced by the network).

The above *MIGHT* be feasible for NU3, but it would likely be at the exclusion of almost all other proposals. In particular, this ZIP would need major revisions in the near future, or NU3 deadlines would need to be extended, in order to be able to fully-specify the necessary changes.